### PR TITLE
Refactoring SteinVI

### PR DESCRIPTION
**Pull request by OlaRonning**
SVGD has been extended to a few varieties ([ASVGD](https://arxiv.org/abs/2101.09815), [GSVGD](https://arxiv.org/abs/2202.03297), [HSVGD](https://arxiv.org/abs/2106.10760)) which I want to include in the einstein module. This PR consists of the changes to `SteinVI` to allow for the extensions and introduction of SVGD and ASVGD. *GSVGD and HSVGD will be in subsequent PRs*.

To this end, I've:
 1. Added a specialized constructor for SVGD. 
    - The SVGD constructor does not require a guide because it's always a delta 
    - It correctly sets the scaling on the attractive force to `1/m` for `m` particles.
    - It doesn't allow users to change `num_elbo_particles` because it's always 1 for SVGD.
 2. Added a `setup_run` method to SteinVI.
    - The method encapsulates a step of SteinVI, which inheriting constructors can manipulate.
 3. Added a constructor for ASVGD.
    -  ASVGD introduces an annealing schedule on the attractive force.
    - This inherits from SVGD and overwrites `setup_run` to change the `loss_temperature`.
 3. Removed [Jacobian projection from the Stein force](https://github.com/pyro-ppl/numpyro/blob/master/numpyro/contrib/einstein/steinvi.py#L337-L339)
    - The projection is unnecessary as we do not allow `AutoIAFNormal`, `AutoBNAFNormal`, `AutoDAIS`, `AutoSemiDAIS` and `AutoSurrogateLikelihoodDAIS`.
    - This simplifies the force computation to attractive+repulsive.

### Misc changes
1. The normalization factor in the `ProbabilityProductKernel` has been removed. The kernel is still a proper kernel; however, this version avoids vanishing/exploding when the guide variances deviate from 1 for "high" dimensional models.
2. Added a rng_key to the kernel. This is convenient when experimenting with kernels.
3. Removed the `enum` parameter as it is currently unsupported.
4. Kernel smoothing of the attractive force and the repulsion is now computed on particles in unconstraint space, consistent with particles moving in unconstraint space. **NB: Constraint particles are removed from the test**.
5. Added the manual computation for kernel tests as comments.
    - I inadvertently changed the value of one of the particles when doing the manual computations, which is why the evaluation has changed.

### TODO
- [x] Add docstrings to ASVGD and SVGD.
- [x] Add manual computation to the IMQKernel and GraphicalKernel tests.
- [x] ~~Fix SteinVI documentation~~ I'll do this in a separate PR.

### CODE REVIEW

File: docs/source/conf.py [hunk 1]
```diff
@@ -226,8 +226,8 @@
 # The theme to use for HTML and HTML Help pages.  See the documentation for
 # a list of builtin themes.
 #
-html_theme = "sphinx_rtd_theme"
-html_theme_path = [sphinx_rtd_theme.get_html_theme_path()]
+# html_theme = "sphinx_rtd_theme"  # <--- FIXME: gives deprecated warning!
```
[the remark this answers is not in the document]
Comment (fehiepsi) [reply docs/source/conf.py:229]: it is fixed upstream. Could you rebase?

[the remark this answers is not in the document]
Comment (OlaRonning) [reply docs/source/conf.py]: Excellent, thanks :+1:

**Comment by OlaRonning** [comment]
@fehiepsi Let me know if you want me to split it into multiple PRs.

File: numpyro/contrib/einstein/mixture_guide_predictive.py [hunk 2]
```diff
@@ -4,21 +4,27 @@
 from collections.abc import Callable, Sequence
 from functools import partial
 from typing import Optional
+import warnings
 
-import jax
 from jax import numpy as jnp, random, vmap
+from jax.tree_util import tree_flatten, tree_map
```
Comment (fehiepsi) [inline numpyro/contrib/einstein/mixture_guide_predictive.py:10]: This is deprecated. You can use `jax.tree.flatten` and `jax.tree.map` now.

File: numpyro/contrib/einstein/steinvi.py [hunk 3]
```diff
@@ -2,30 +2,25 @@
...
 import functools
 from functools import partial
 from itertools import chain
 import operator
 
-import jax
-from jax import grad, jacfwd, numpy as jnp, random, vmap
+from jax import grad, numpy as jnp, random, vmap
 from jax.flatten_util import ravel_pytree
+from jax.tree_util import tree_map
```
Comment (fehiepsi) [inline numpyro/contrib/einstein/steinvi.py:13]: same as above

File: test/contrib/einstein/test_steinvi_util.py [hunk 4]
```diff
@@ -8,8 +8,8 @@
 import pytest
 import scipy
 
-import jax
 from jax import numpy as jnp
+from jax.tree_util import tree_flatten, tree_map
```
Comment (fehiepsi) [inline test/contrib/einstein/test_steinvi_util.py:12]: same as above

**Review by fehiepsi — APPROVED** [review]
LGTM pending minor comments.
